### DISCUSSION TITLE
`clap_plugin_factory` functions take itself as a parameter?

### DISCUSSION THREAD

**Discussion by rokit** [free-audio/clap#243] [Q&A] [answered]
Why does [`clap_plugin_factory`](https://github.com/free-audio/clap/blob/84531b931c12285097746a9cae77690a680b8aa0/include/clap/plugin-factory.h#L16) have functions that also take `clap_plugin_factory` as a parameter? The parameter is not used in the [template](https://github.com/free-audio/clap/blob/84531b931c12285097746a9cae77690a680b8aa0/src/plugin-template.c#L298). For example, why not have a `uint32_t plugin_count` member instead of a method?

**Comment by defiantnerd** [comment, marked as answer]
It is possible to have several factories and you might need to address the factory, imagine it as a "this" pointer. The parameter is not used in the template since it has only one factory, but more complex plugin libraries are possible where the appropriate factory needs to be selected.

In general it is also useful not to have a simple data member but a function which can determine things at runtime (like licenses, dynamic wrappers etc.). This way you can set up the structs as static/const and keep the flexibility without the necessity of code that just prepares structs.
